### DISCUSSION TITLE
Why `Value` is `pub` while its bound `Sealed` is sealed?

### DISCUSSION THREAD

**Discussion by tisonkun** [tokio-rs/tracing#1455] [Q&A] [answered]
```rust
pub trait Value: crate::sealed::Sealed {
    /// Visits this value with the given `Visitor`.
    fn record(&self, key: &Field, visitor: &mut dyn Visit);
}

mod sealed {
    pub trait Sealed {}
}
```

I don't see a solution to implement Value in application code. I can see that it is possible using `DebugValue` or other out-of-the-box implementation for boxing the value, but why `Value` is `pub` while its bound `Sealed` is sealed?

**Comment by hawkw** [comment, marked as answer]
`Value` is not intended to be implemented in user code, that's why it's `Sealed`. It's only implemented for certain Rust primitive types. The intention is to replace the current value system with a more sophisticated implementation that allows user code to implement `Value` for arbitrary types,  but this will require some significant changes to the trait. Until then, the trait has been sealed to prevent downstream implementations that would be broken by those changes.

However, the trait must be _used_ in user code. Therefore, it has to be `pub`. `pub` traits bounded with `Sealed` allow traits to be _used_ publicly while preventing downstream implementations. See https://rust-lang.github.io/api-guidelines/future-proofing.html#sealed-traits-protect-against-downstream-implementations-c-sealed

**Reply by tisonkun** [reply]
Thanks for your detailed answer and I just notice that we have a Discussion forum.

**Reply by rollo-b2c2** [reply]
Is there any reason this is still sealed?  It's kinda annoying to be honest, if it breaks it breaks.

I'm happy to pin my tracing version.  It has all the features I want.  If there's a new feature and it breaks I can weigh up if the time invested in porting it.

**Reply by davidbarsky** [reply]
Because the successor to `Value` is `valuable`. You can enable it today, if you'd like: https://docs.rs/tracing/latest/tracing/#unstable-features.
